### PR TITLE
MODELIX-524 unstable ReplicatedRepositoryTest

### PR DESCRIPTION
**Pull request by slisson**
The reason that the test failed sometimes was that each replica kept creating new merge commits after receiving an update, causing the history to grow without any actual change to the model. The complexity of the merge operation also kept growing and at some point the algorithm just took to long and the synchronization didn't finish within the time limit.

### CODE REVIEW

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt [hunk 1]
```diff
@@ -147,10 +149,18 @@ class RepositoriesManager(val client: LocalModelClient) {
...
     }
 
+    private val deltaCache = LRUMap<Pair<String, String?>, Lazy<Map<String, String>>>(10)
     fun computeDelta(versionHash: String, baseVersionHash: String?): Map<String, String> {
-        val version = CLVersion(versionHash, client.storeCache)
-        val baseVersion = baseVersionHash?.let { CLVersion(it, client.storeCache) }
-        return version.computeDelta(baseVersion)
+        return runSynchronized(deltaCache) {
+            deltaCache.getOrPut(versionHash to baseVersionHash) {
+                // lazy { ... } allows to run the computation without locking deltaCache
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt:156]: Can there be a race condition that the cache is invalidated once the computation actually starts?

File: model-datastructure/src/commonMain/kotlin/org/modelix/model/LinearHistory.kt [hunk 2]
```diff
@@ -5,6 +5,10 @@ import org.modelix.model.lazy.IDeserializingKeyValueStore
 import org.modelix.model.lazy.KVEntryReference
 import org.modelix.model.persistent.CPVersion
 
+/**
+ * When merging two branches there might be conflicts between operations that need to be resolved in a deterministic
+ * way. This class puts them in an order that preserves the order of earlier merges.
```
Comment (languitar) [inline model-datastructure/src/commonMain/kotlin/org/modelix/model/LinearHistory.kt:10]: Can you give some ideas or references for the algorithm that achieves this property?

File: model-datastructure/src/commonMain/kotlin/org/modelix/model/operations/OTBranch.kt [hunk 3]
```diff
@@ -30,30 +30,47 @@ class OTBranch(
...
-                operations = ArrayList()
-                return Pair(newOperations, treeForOperations)
+            return runSynchronized(completedChanges) {
+                val result = when (completedChanges.size) {
+                    0 -> emptyList<IAppliedOperation>() to computeReadT { it.tree }
+                    1 -> completedChanges[0]
+                    else -> completedChanges.flatMap { it.first } to completedChanges.last().second
+                }
+                completedChanges.clear()
+                result
```
Comment (languitar) [inline model-datastructure/src/commonMain/kotlin/org/modelix/model/operations/OTBranch.kt:56]: This is the same implementation as in `getPendingChanges`. Can't this getter redirect to that method to avoid the duplication?

File: model-datastructure/src/commonMain/kotlin/org/modelix/model/operations/OTBranch.kt [hunk 4]
```diff
@@ -30,30 +30,47 @@ class OTBranch(
...
+
+    /**
+     * @return the operations applied to the branch since the last call of this function and the resulting ITree.
+     */
+    fun getPendingChanges(): Pair<List<IAppliedOperation>, ITree> {
+        return runSynchronized(completedChanges) {
+            val result = when (completedChanges.size) {
+                0 -> emptyList<IAppliedOperation>() to computeReadT { it.tree }
+                1 -> completedChanges[0]
+                else -> completedChanges.flatMap { it.first } to completedChanges.last().second
```
Comment (languitar) [inline model-datastructure/src/commonMain/kotlin/org/modelix/model/operations/OTBranch.kt:68]: That's a bit hard to read. Maybe an explanatory comment helps the next reader here?

File: model-datastructure/src/commonMain/kotlin/org/modelix/model/operations/OTBranch.kt [hunk 5]
```diff
@@ -78,7 +95,21 @@ class OTBranch(
...
+                    val result = computable()
+                    runSynchronized(completedChanges) {
+                        completedChanges += currentOperations to t.tree
+                    }
+                    result
+                } finally {
+                    currentOperations = ArrayList()
+                }
+            }
+        }
```
Comment (languitar) [inline model-datastructure/src/commonMain/kotlin/org/modelix/model/operations/OTBranch.kt:112]: Can you add some explanations on the two conditions here? They are hard to understand without prior knowledge.

File: model-server/build.gradle.kts [hunk 6]
```diff
@@ -107,7 +107,7 @@ val cucumber = task("cucumber") {
     }
 }
 
-tasks.named("test") {
+tasks.named("build") {
```
Comment (languitar) [inline model-server/build.gradle.kts:110]: Would it make sense to migrate this setup to [test suites](https://blog.gradle.org/introducing-test-suites)?

File: model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt [hunk 7]
```diff
@@ -45,11 +45,32 @@ class VersionMerger(private val storeCache: IDeserializingKeyValueStore, private
...
         val commonBase = Companion.commonBaseVersion(leftVersion, rightVersion)
         if (commonBase?.hash == leftVersion.hash) return rightVersion
         if (commonBase?.hash == rightVersion.hash) return leftVersion
+
+        val leftNonMerges = HashSet<Long>().also { collectLatestNonMerges(leftVersion, HashSet(), it) }
+        val rightNonMerges = HashSet<Long>().also { collectLatestNonMerges(rightVersion, HashSet(), it) }
+        if (leftNonMerges == rightNonMerges) {
+            // If there is no actual change on both sides, but they just did the same merge, we have to pick one
+            // of them, otherwise both sides will continue creating merges forever.
+            return if (leftVersion.id < rightVersion.id) leftVersion else rightVersion
```
Comment (languitar) [inline model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt:86]: Maybe a detailed and easier to understand unit test for exactly this condition and function would be a good addition to the test code base. The ReplicatedRepositoryTest is pretty complex and maybe there's a more targeted version that also helps documenting the corner-case for others reading the code.

File: model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt [hunk 8]
```diff
@@ -45,11 +45,32 @@ class VersionMerger(private val storeCache: IDeserializingKeyValueStore, private
...
+    private fun collectLatestNonMerges(version: CLVersion?, visited: MutableSet<String>, result: MutableSet<Long>) {
+        if (version == null) return
+        if (!visited.add(version.getContentHash())) return
+        if (version.isMerge()) {
+            collectLatestNonMerges(version.getMergedVersion1(), visited, result)
+            collectLatestNonMerges(version.getMergedVersion2(), visited, result)
+        } else {
+            result.add(version.id)
+            collectLatestNonMerges(version.baseVersion, visited, result)
+        }
```
Comment (languitar) [inline model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt:57]: How deeply nested is this recursion usually? Is there a chance for reaching a stack overflow before this converges?

File: model-server/src/test/kotlin/org/modelix/model/server/ReplicatedRepositoryTest.kt [hunk 9]
```diff
@@ -116,9 +115,27 @@ class ReplicatedRepositoryTest {
...
+
+    @Test fun `concurrent write 6`() = concurrentWrite(6)
+
+    @Test fun `concurrent write 7`() = concurrentWrite(7)
+
+    @Test fun `concurrent write 8`() = concurrentWrite(8)
+
+    @Test fun `concurrent write 9`() = concurrentWrite(9)
+
+    fun concurrentWrite(iteration: Int) = runTest {
```
Comment (languitar) [inline model-server/src/test/kotlin/org/modelix/model/server/ReplicatedRepositoryTest.kt]: This setup can be simplified with [`@RepeatedTest`](https://www.baeldung.com/junit-5-repeated-test#accessing-the-repetitioninfo).

File: model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt [hunk 10]
```diff
@@ -2,68 +2,61 @@ package org.modelix.model.client2
...
-import kotlin.coroutines.cancellation.CancellationException
 
-class ReplicatedModel(val client: IModelClientV2, val branchRef: BranchReference, val query: ModelQuery? = null) {
-    private val scope = CoroutineScope(Dispatchers.Default)
+class ReplicatedModel(
+    val client: IModelClientV2,
+    val branchRef: BranchReference,
+    private val providedScope: CoroutineScope? = null,
+    @Deprecated("was never supported")
+    val query: ModelQuery? = null,
```
Comment (languitar) [inline model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt:30]: Is there anyone client that actually used this parameter if it didn't work? If not, we could avoid the deprecation at all and just drop it. I would assume only modelix-internal clients of this class?

File: model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt [hunk 11]
```diff
@@ -73,79 +66,92 @@ class ReplicatedModel(val client: IModelClientV2, val branchRef: BranchReference
...
                 scope.launch {
                     pushLocalChanges()
+//                    while (state != State.Disposed) {
+//                        if (remoteVersion.getNumberOfUnconfirmed() == 0) {
+//                            pushLocalChanges()
+//                            break
+//                        } else {
+//                            delay(100.milliseconds)
+//                        }
+//                    }
```
Comment (languitar) [inline model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt:99]: Why is this commented?

File: model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt [hunk 12]
```diff
@@ -73,79 +66,92 @@ class ReplicatedModel(val client: IModelClientV2, val branchRef: BranchReference
...
+//                    remoteVersionReceived(newRemoteVersion)
+//                    nextDelayMs = 1000
+//                } catch (ex: CancellationException) {
+//                    break
+//                } catch (ex: Throwable) {
+//                    LOG.error(ex) { "Failed to pull branch $branchRef" }
+//                    nextDelayMs = (nextDelayMs * 3 / 2).coerceIn(1000, 30000)
+//                }
+//            }
+//        }
```
Comment (languitar) [inline model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt:85]: Can we put this into a separate method to avoid polluting the production code with commented-out debugging aids?

File: model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt [hunk 13]
```diff
@@ -63,6 +63,21 @@ class VersionMerger(private val storeCache: IDeserializingKeyValueStore, private
...
+//
+//        if (leftNonMerges == rightNonMerges) {
+//            // If there is no actual change on both sides, but they just did the same merge, we have to pick one
+//            // of them, otherwise both sides will continue creating merges forever.
+//            return if (leftVersion.id < rightVersion.id) leftVersion else rightVersion
+//        } else if ((leftNonMerges - rightNonMerges).isEmpty()) {
+//            return rightVersion // the right history already contains all the left history
+//        } else if ((rightNonMerges - leftNonMerges).isEmpty()) {
+//            return leftVersion // the left history already contains all the right history
+//        }
```
Comment (languitar) [inline model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt:79]: Why commented out? Does the history of this PR needs some rebasing for cleanup?

File: model-server/src/test/kotlin/org/modelix/model/server/ReplicatedRepositoryTest.kt [hunk 14]
```diff
@@ -63,7 +72,10 @@ class ReplicatedRepositoryTest {
             install(WebSockets)
             ModelReplicationServer(InMemoryStoreClient()).init(this)
         }
-        block()
+
+        coroutineScope {
+            block(this)
+        }
```
Comment (languitar) [inline model-server/src/test/kotlin/org/modelix/model/server/ReplicatedRepositoryTest.kt:78]: Can this be replaced with [`runTest`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/run-test.html)?

**Review by languitar — CHANGES_REQUESTED** [review]
The algorithmic changes are pretty hard to review for me because the commit "fix(model-client): fixed unstable ReplicatedRepositoryTest", which seems to do the major work, is pretty large and lacks comments to explain what happened. Moreover, that commit comments out some change that were introduced before. Is there a chance to clean up the history and to provide a few more information regarding these changes?

Apart from that, I have found a few spots where minor things could be improved or might need some clarifications.

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt (hunk 1 above)
Comment (slisson) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt:156]: The input is immutable. It never gets invalid.

File: model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt (hunk 7 above)
Comment (slisson) [reply model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt:86]: Maybe in a separate PR.

File: model-server/build.gradle.kts (hunk 6 above)
Comment (slisson) [reply model-server/build.gradle.kts:114]: Maybe in a separate PR.
